ckanext/storepublisher/controllers/ui_controller.py

```diff
@@ -56,42 +79,119 @@ def publish(self, id, offering_info=None, errors=None):
+                    c.user, id))
+            tk.abort(
+                401, tk._('User %s not authorized to publish %s') % (
+                    c.user, id))
 
         # Get the dataset and set template variables
         # It's assumed that the user can view a package if he/she can update it
+
         dataset = tk.get_action('package_show')(context, {'id': id})
         c.pkg_dict = dataset
         c.errors = {}
 
-        # Tag string is needed in order to set the list of tags in the form
-        if 'tag_string' not in c.pkg_dict:
-            tags = [tag['name'] for tag in c.pkg_dict.get('tags', [])]
-            c.pkg_dict['tag_string'] = ','.join(tags)
-
+        def _sort_categories(categories):
+            listOfTags = []
+            catRelatives = {}
+            tagSorted = sorted(categories, key=lambda x: int(x['id']))
+            if not len(tagSorted):
+                return listOfTags, catRelatives
+            listOfTags.append(tagSorted[0])
+            catRelatives[tagSorted[0]['id']] = {'href': tagSorted[0]['href'],
+                                                'id': tagSorted[0]['id']}
+            tagSorted.pop(0)
+
+            # Im sorry for this double loop, ill try to optimize this
+            for tag in tagSorted:
+                if tag['isRoot']:
+                    listOfTags.append(tag)
+                    catRelatives[tag['id']] = {'href': tag['href'],
+                                               'id': tag['id']}
+                    continue
+                for item in listOfTags:
+                    if tag['parentId'] == item['id']:
+                        listOfTags.insert(listOfTags.index(item) + 1, tag)
+                        catRelatives[tag['id']] = {'href': tag['href'],
+                                                   'id': tag['id'],
+                                                   'parentId': tag.get('parentId', '')}
+                        break
+            return listOfTags, catRelatives
+
+        # Get tags in the expected format of the form select field
+        def _getList(param):
+            requiredFields = ['id', 'name']
+            result = []
+            for i in param:
+                result.append({x: i[x] for x in requiredFields})
+            for elem in result:
+                elem['text'] = elem.pop('name')
+                elem['value'] = elem.pop('id')
+            return result
+
+        def _validateVersion(version):
+            ver = version
+            if not ver:
+                ver = '1.0'
+            if re.search(r'\.$', ver) is not None:
+                ver += '0'
+            if re.search(r'\.{2,}', ver) is not None:
+                ver = re.sub(r'\.{2,}', r'\.', ver)
+            if re.search(r'^\.', ver) is not None:
+                ver = "1" + ver
+            return ver
+
+        if request.GET:
+            self._listOfTags, self._catRelatives = _sort_categories(self._get_Content('category'))
```

Comment:
Review the name of the vars. In pyhton it is not used camel case for methods and variables. Make a general review on this